testcases/kernel/syscalls/open/open09.c

```diff
@@ -1,109 +1,57 @@
- *   but WITHOUT ANY WARRANTY;  without even the implied warranty of
- *   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See
- *   the GNU General Public License for more details.
- *
- *   You should have received a copy of the GNU General Public License
- *   along with this program;  if not, write to the Free Software
- *   Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA
- */
+// SPDX-License-Identifier: GPL-2.0-or-later
 
 /*
- * Description:
- *	1. open an O_WRONLY file, test if read failed.
- *	2. open an O_RDONLY file, test if write failed.
+ * Copyright (c) International Business Machines  Corp., 2002
+ * Copyright (c) 2013 Wanlong Gao <gaowanlong@cn.fujitsu.com>
+ * Copyright (C) 2024 SUSE LLC Andrea Manzini <andrea.manzini@suse.com>
  */
 
-#include <sys/types.h>
-#include <sys/stat.h>
-#include <stdio.h>
-#include <fcntl.h>
-#include "test.h"
-
-char *TCID = "open09";
-int TST_TOTAL = 2;
-
-#define PASSED 1
-#define FAILED 0
+/*\
+ * [Description]
+ * 1. open an O_RDONLY file, test if write failed.
+ * 2. open an O_WRONLY file, test if read failed.
+ */
 
-static char tempfile[40] = "";
+#include "tst_test.h"
 
-static void setup(void);
-static void cleanup(void);
+#define TEMPFILE "testfile"
 
-int main(int ac, char *av[])
+static void setup(void)
 {
-	int fildes;
-	int ret = 0;
-	char pbuf[BUFSIZ];
-
-	int lc;
-
-	tst_parse_opts(ac, av, NULL, NULL);
-
-	setup();
-
-	for (lc = 0; TEST_LOOPING(lc); lc++) {
-
-		fildes = open(tempfile, O_WRONLY);
-		if (fildes == -1)
-			tst_brkm(TFAIL, cleanup, "\t\t\topen failed");
-
-		ret = read(fildes, pbuf, 1);
-		if (ret != -1)
-			tst_resm(TFAIL, "read should not succeed");
-		else
-			tst_resm(TPASS, "Test passed in O_WRONLY.");
-
-		close(fildes);
-
-		fildes = open(tempfile, O_RDONLY);
-		if (fildes == -1) {
-			tst_resm(TFAIL, "\t\t\topen failed");
-		} else {
-			ret = write(fildes, pbuf, 1);
-			if (ret != -1)
-				tst_resm(TFAIL, "write should not succeed");
-			else
-				tst_resm(TPASS, "Test passed in O_RDONLY.");
-		}
-		close(fildes);
-	}
-
-	cleanup();
-	tst_exit();
+	SAFE_CREAT(TEMPFILE, 0600);
 }
 
-static void setup(void)
+static void cleanup(void)
 {
-	int fildes;
-
-	tst_sig(NOFORK, DEF_HANDLER, cleanup);
-	TEST_PAUSE;
-	tst_tmpdir();
-
-	sprintf(tempfile, "open09.%d", getpid());
-
-	fildes = creat(tempfile, 0600);
-	if (fildes == -1) {
-		tst_brkm(TBROK, cleanup, "\t\t\tcan't create '%s'",
-			 tempfile);
-	} else {
-		close(fildes);
-	}
+	SAFE_UNLINK(TEMPFILE);
 }
 
-static void cleanup(void)
+static void verify_open(unsigned int nr)
 {
-	unlink(tempfile);
-	tst_rmdir();
+	char pbuf[BUFSIZ];
+	int fd = 0;
+
+	switch (nr)	{
+	case 0:
+		fd = SAFE_OPEN(TEMPFILE, O_RDONLY);
+		TST_EXP_FAIL(write(fd, pbuf, 1), EBADF);
+		break;
+	case 1:
+		fd = SAFE_OPEN(TEMPFILE, O_WRONLY);
+		TST_EXP_FAIL(read(fd, pbuf, 1), EBADF);
+		break;
+	default:
+		// should never reach this
+		tst_brk(TBROK, "Invalid testcase");
+		break;
```

Comment:
Better to use `if else` syntax because this looks like a `switch/case` on a boolean..